Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2021, Ali Mohammad Pur <mpfard@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/Array.h>
+#include <LibJS/Runtime/ArrayBuffer.h>
+#include <LibJS/Runtime/BigInt.h>
+#include <LibJS/Runtime/TypedArray.h>
+#include <LibWasm/AbstractMachine/Interpreter.h>
+#include <LibWeb/WebAssembly/WebAssemblyObject.h>
+
+namespace Web::Bindings {
+
+WebAssemblyObject::WebAssemblyObject(JS::GlobalObject& global_object)
+    : Object(*global_object.object_prototype())
+{
+}
+
+void WebAssemblyObject::initialize(JS::GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+
+    define_native_function("validate", validate, 1);
+    define_native_function("compile", compile, 1);
+    define_native_function("instantiate", instantiate, 1);
+}
+
+NonnullOwnPtrVector<WebAssemblyObject::CompiledWebAssemblyModule> WebAssemblyObject::s_compiled_modules;
+NonnullOwnPtrVector<Wasm::ModuleInstance> WebAssemblyObject::s_instantiated_modules;
+Wasm::AbstractMachine WebAssemblyObject::s_abstract_machine;
+
+JS_DEFINE_NATIVE_FUNCTION(WebAssemblyObject::validate)
+{
+    // FIXME: Implement this once module validation is implemented in LibWasm.
+    dbgln("Hit WebAssemblyObject::validate() stub!");
+    return JS::Value { true };
+}
+
+static Result<size_t, JS::Value> parse_module(JS::GlobalObject& global_object, JS::Object* buffer)
+{
+    ByteBuffer bytes;
+    if (is<JS::ArrayBuffer>(buffer)) {
+        auto array_buffer = static_cast<JS::ArrayBuffer*>(buffer);
+        bytes = array_buffer->buffer();
+    } else if (is<JS::TypedArrayBase>(buffer)) {
+        auto array = static_cast<JS::TypedArrayBase*>(buffer);
+        bytes = array->viewed_array_buffer()->buffer();
+    } else {
+        auto error = JS::TypeError::create(global_object, String::formatted("{} is not an ArrayBuffer", buffer->class_name()));
```

Comment:
That formatting capability is only available via `VM::throw_exception(GlobalObject&, ErrorType, Args...)`, the individual error classes do not have it.

Reply:
oh, right. Unfortunate, but nevermind then.

Reply:
I could add one, if you provide a name for it that doesn't sound terrible like `JS::Error::create_error<JS::TypeError>(...)`

Reply:
I think ideally it would just be an overload of `ErrorType::create()` if that's possible? I.e. either `TypeError::create(global_object, ErrorType,  Args...)` or `TypeError::create(global_object, const String&)`. Let's not get carried away though, we can deal with this separately.

Reply:
Alright, we'll add it later then : )